hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java

```diff
@@ -0,0 +1,406 @@
+        });
+    }
+
+    /**
+     * @param source A {@link BatchSource BatchSource&lt;Long&gt;} or {@link StreamSource
+     *        StreamSource&lt;Long&gt;} where the first element is {@code 0L}, which is used
+     *        to increment the {@linkplain JobStatusLogger#runCount run count} of the job.
+     */
+    protected void testListener(JobConfig config, Object source, BiConsumer<Job, JobStatusLogger> test) {
+        Pipeline p = Pipeline.create();
+        int jobId = nextJobId++;
+        (source instanceof BatchSource
+                    ? p.readFrom((BatchSource<Long>) source)
+                    : p.readFrom((StreamSource<Long>) source).withoutTimestamps())
+                .writeTo(Sinks.<Long, Integer, Integer>mapWithUpdating(
+                        MAP_NAME, s -> jobId, (i, s) -> s == 0 ? (i == null ? 0 : i) + 1 : i));
+
+        Job job = instance.get().getJet().newJob(p, config);
+        JobStatusLogger listener = new JobStatusLogger(job, jobId);
+        jobIdString = job.getIdString();
+        registrationId = listener.registrationId;
+        test.accept(job, listener);
+    }
+
+    protected void testListener(Object source, BiConsumer<Job, JobStatusLogger> test) {
+        testListener(new JobConfig(), source, test);
+    }
+
+    protected void testListener(Object source, Consumer<Job> test, String... log) {
+        testListener(source, (job, listener) -> {
+            test.accept(job);
+            assertTailEqualsEventually(listener.log, log);
+        });
+    }
+
+    protected void testLightListener(Object source, BiConsumer<Job, JobStatusLogger> test) {
+        Pipeline p = Pipeline.create();
+        (source instanceof BatchSource
+                    ? p.readFrom((BatchSource<?>) source)
+                    : p.readFrom((StreamSource<?>) source).withoutTimestamps())
+                .writeTo(Sinks.noop());
+
+        Job job = instance.get().getJet().newLightJob(p);
+        JobStatusLogger listener = new JobStatusLogger(job, -1);
+        jobIdString = job.getIdString();
+        registrationId = listener.registrationId;
+        test.accept(job, listener);
+    }
+
+    protected void testLightListener(Object source, Consumer<Job> test, String log) {
+        testLightListener(source, (job, listener) -> {
+            test.accept(job);
+            assertIterableEqualsEventually(listener.log, log);
+        });
+    }
+
+    protected static void assertIterableEqualsEventually(Iterable<?> actual, Object... expected) {
+        assertTrueEventually(() -> assertIterableEquals(actual, expected));
+    }
+
+    @SafeVarargs
+    protected static <T> void assertTailEqualsEventually(List<T> actual, T... expected) {
+        assertTrueEventually(() -> {
+            assertBetween("length", actual.size(), expected.length - 2, expected.length);
+            List<T> tail = asList(expected).subList(expected.length - actual.size(), expected.length);
```

Comment:
listener.log changes in parallel with this check - can cause `java.util.ConcurrentModificationException`